pkg/agent/agent.go

```diff
@@ -226,50 +226,94 @@ func (i *Initializer) initInterfaceStore() error {
+			// Set the gateway interface name to the discovered name.
+			i.hostGateway = intf.InterfaceName
+		}
+		return intf
+	}
+	parseUplinkIFFunc := func(port *ovsconfig.OVSPortData, ovsPort *interfacestore.OVSPortConfig) *interfacestore.InterfaceConfig {
+		return &interfacestore.InterfaceConfig{
+			Type:          interfacestore.UplinkInterface,
+			InterfaceName: port.Name,
+			OVSPortConfig: ovsPort,
+		}
+	}
+	parseTunnelIFFunc := func(port *ovsconfig.OVSPortData, ovsPort *interfacestore.OVSPortConfig) *interfacestore.InterfaceConfig {
+		intf := noderoute.ParseTunnelInterfaceConfig(port, ovsPort)
+		if intf != nil && port.OFPort == config.DefaultTunOFPort &&
+			intf.InterfaceName != i.nodeConfig.DefaultTunName {
+			klog.Infof("The discovered default tunnel interface name %s is different from the default value: %s",
+				intf.InterfaceName, i.nodeConfig.DefaultTunName)
+			// Set the default tunnel interface name to the discovered name.
+			i.nodeConfig.DefaultTunName = intf.InterfaceName
+		}
+		return intf
+	}
 	ifaceList := make([]*interfacestore.InterfaceConfig, 0, len(ovsPorts))
-	uplinkIfName := i.nodeConfig.UplinkNetConfig.Name
 	for index := range ovsPorts {
 		port := &ovsPorts[index]
 		ovsPort := &interfacestore.OVSPortConfig{
 			PortUUID: port.UUID,
 			OFPort:   port.OFPort}
 		var intf *interfacestore.InterfaceConfig
-		switch {
-		case port.OFPort == config.HostGatewayOFPort:
-			intf = &interfacestore.InterfaceConfig{
-				Type:          interfacestore.GatewayInterface,
-				InterfaceName: port.Name,
-				OVSPortConfig: ovsPort}
-			if intf.InterfaceName != i.hostGateway {
-				klog.Warningf("The discovered gateway interface name %s is different from the configured value: %s",
-					intf.InterfaceName, i.hostGateway)
-				// Set the gateway interface name to the discovered name.
-				i.hostGateway = intf.InterfaceName
+		if port.AntreaIFType != ovsconfig.AntreaUnSet {
+			switch port.AntreaIFType {
+			case ovsconfig.AntreaGateway:
+				intf = parseGatewayIFFunc(port, ovsPort)
+			case ovsconfig.AntreaUplink:
+				intf = parseUplinkIFFunc(port, ovsPort)
+			case ovsconfig.AntreaTunnel:
+				intf = parseTunnelIFFunc(port, ovsPort)
+			case ovsconfig.AntreaHostIF:
+				// Not load the host interface, because it is configured on the OVS bridge port, and we don't need a
+				// specific interface in the interfaceStore.
+				intf = nil
+			case ovsconfig.AntreaContainer:
+				// The port should be for a container interface.
+				intf = cniserver.ParseOVSPortInterfaceConfig(port, ovsPort, true)
+			default:
+				klog.InfoS("Unknown Antrea Interface type", "type", port.AntreaIFType)
 			}
-		case port.Name == uplinkIfName:
-			intf = &interfacestore.InterfaceConfig{
-				Type:          interfacestore.UplinkInterface,
-				InterfaceName: port.Name,
-				OVSPortConfig: ovsPort,
+		} else {
+			// Antrea Interface type is not saved in OVS Interface external_ids in earlier Antrea versions, so we use
```

Comment:
Since this change will affect the existing storage and introduces upgrade complexity, I'd like to understand why we need these types explicitly stored. Doesn't the "type" of OVS interface already differentiate internal port, tunnel port, and attached port (including container port and uplink port)? What do we want to differentiate further?

Reply:
The existing code is actually using the predefined ofport number to parse some kinds of the interfaces, e.g. gw0, uplink, this has two weaks: 1) If the OF port number is taken before creating the OVS port, the actual port number allocated by OVS is different from ofport_request, then we might fail in restart case when using these predefined numbers; 2) It is difficult to expand to multiple interfaces with such types, like supporting multiple uplink interfaces. Using a fixed type value can resolve the above issues.